File: tutorial-1-translating-with-jobs/java/src/main/java/com/smartling/samples/jobstutorial/ProcessJob.java
package com.smartling.samples.jobstutorial;

import com.smartling.api.files.v2.FilesApi;
import com.smartling.api.files.v2.FilesApiFactory;
import com.smartling.api.files.v2.pto.DownloadTranslationPTO;
import com.smartling.api.files.v2.pto.RetrievalType;
import com.smartling.api.jobbatches.v2.JobBatchesApi;
import com.smartling.api.jobbatches.v2.JobBatchesApiFactory;
import com.smartling.api.jobbatches.v2.pto.CreateBatchRequestPTO;
import com.smartling.api.jobbatches.v2.pto.CreateBatchResponsePTO;
import com.smartling.api.jobbatches.v2.pto.StreamFileUploadPTO;
import com.smartling.api.jobs.v3.TranslationJobsApi;
import com.smartling.api.jobs.v3.TranslationJobsApiFactory;
import com.smartling.api.jobs.v3.pto.ContentProgressReportPTO;
import com.smartling.api.jobs.v3.pto.TranslationJobCreateCommandPTO;
import com.smartling.api.jobs.v3.pto.TranslationJobCreateResponsePTO;
import com.smartling.api.v2.authentication.AuthenticationApi;
import com.smartling.api.v2.authentication.AuthenticationApiFactory;
import com.smartling.api.v2.client.ClientFactory;
import com.smartling.api.v2.client.DefaultClientConfiguration;
import com.smartling.api.v2.client.auth.Authenticator;
import com.smartling.api.v2.client.auth.BearerAuthSecretFilter;
import com.smartling.api.v2.client.exception.RestApiRuntimeException;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.*;
import java.util.concurrent.TimeUnit;


public class ProcessJob
{
    private static final List<String> LOCALES = new ArrayList<>(Arrays.asList("fr-FR"));
    private static final List<String> FILE_LIST = new ArrayList<>(Arrays.asList("site-navigation.json","products.json"));
    private static final String FILE_TYPE = "json";

    public static void main( String[] args ) throws Exception
    {
        // Read command-line arguments
        if (args.length != 1) {
            System.out.println("Usage: mvn exec:java@run");
            System.out.println("(See pom.xml to configure parameters.)");
            System.exit(1);
        }
        String jobName = args[0];

        // Read authentication credentials from environment
        String userId = System.getenv("DEV_USER_IDENTIFIER");
        String userSecret = System.getenv("DEV_USER_SECRET");
        String projectId = System.getenv("DEV_PROJECT_ID");
        if (userId == null || userSecret == null || projectId == null) {
            System.out.println("Missing environment variables. Did you run setenv?");
            System.exit(1);
        }

        // Create factory for building API clients
        DefaultClientConfiguration clientConfiguration = DefaultClientConfiguration.builder().build();
        ClientFactory clientFactory = new ClientFactory();

        // Create the authenticator to be used by subsequent API calls
        AuthenticationApi authenticationApi = new AuthenticationApiFactory(clientFactory).buildApi(clientConfiguration);
        Authenticator authenticator = new Authenticator(userId, userSecret, authenticationApi);
        BearerAuthSecretFilter bearerAuthSecretFilter = new BearerAuthSecretFilter(authenticator);

        // Instantiate the APIs we'll need
        TranslationJobsApi jobsApi = new TranslationJobsApiFactory(clientFactory)
                .buildApi(bearerAuthSecretFilter, clientConfiguration);
        JobBatchesApi jobBatchesApi = new JobBatchesApiFactory(clientFactory)
                .buildApi(bearerAuthSecretFilter, clientConfiguration);
        FilesApi filesApi = new FilesApiFactory(clientFactory)
                .buildApi(bearerAuthSecretFilter, clientConfiguration);


        try {
            System.out.println();

            // Create job
            System.out.println("Calling 'create job' endpoint...");
            String jobUid = createJob(jobsApi, projectId, jobName);
            System.out.println("Job created. Job ID: " + jobUid);

            // Generate URIs for the test files
            // For testing purposes, we're prepending the job ID to the file names to generate a unique
            // URI for each run of the test, thus allowing the same file to be added to different jobs.
            Map<String, String> urisByFilename = new HashMap<>();
            for (String fileName : FILE_LIST) {
                String uri = jobUid + "/" + fileName;
                urisByFilename.put(fileName, uri);
            }

            // Create job batch
            System.out.println("Calling 'create job batch' endpoint...");
            List<String> fileUris = new ArrayList<>(urisByFilename.values());
            String batchUid = createJobBatch(jobBatchesApi, projectId, jobUid, fileUris);
            System.out.println("Job batch created. Batch ID: " + batchUid);

            // Add files to batch
            System.out.println("Calling 'upload file to batch' endpoint...");
            for (String fileName : FILE_LIST) {
                String fileUri = urisByFilename.get(fileName);
                addFileToBatch(jobBatchesApi, projectId, batchUid, fileName, fileUri);
                System.out.println("Added file " + fileName);
            }
            System.out.println("Finished adding files to batch");

            // Check job status until complete
            System.out.println("Will check job status every 10 seconds");
            int percentComplete = 0;
            do {
                TimeUnit.SECONDS.sleep(10);
                ContentProgressReportPTO response = jobsApi.getTranslationJobProgress(projectId, jobUid, "");
                percentComplete = response.getProgress().getPercentComplete();
                System.out.println("Job Progress: " + percentComplete  + "%");
            }
            while (percentComplete < 100);
            System.out.println("Job is complete; downloading translated files...");

            // Download translated files
            for (String fileName : FILE_LIST) {
                String fileUri = urisByFilename.get(fileName);
                for (String locale : LOCALES) {
                    InputStream translations = downloadTranslatedFile(filesApi, projectId, fileUri, locale);
                    String outputFileName = getOutputFileName(fileName, locale);
                    File outputFile = new File(outputFileName);
                    FileUtils.copyInputStreamToFile(translations, outputFile);
                    System.out.println("Downloaded " + outputFileName);
                }
            }

        } catch (RestApiRuntimeException e) {
            System.err.println("Request error. Response details:");
            System.err.println(e.getMessage());
            System.err.println();
        }

    }

    public static String createJob(TranslationJobsApi jobsApi, String projectId, String jobName) throws RestApiRuntimeException
    {
        TranslationJobCreateCommandPTO requestBody = TranslationJobCreateCommandPTO.builder()
                .jobName(jobName)
                .build();

        TranslationJobCreateResponsePTO response = jobsApi.createTranslationJob(projectId, requestBody);

        return response.getTranslationJobUid();
    }


    public static String createJobBatch(JobBatchesApi jobBatchesApi, String projectId, String jobUid, List fileUris) throws RestApiRuntimeException
    {
        CreateBatchRequestPTO requestBody = CreateBatchRequestPTO.builder()
                .translationJobUid(jobUid)
                .authorize(true)
                .fileUris(fileUris)
                .build();

        CreateBatchResponsePTO response = jobBatchesApi.createBatch(projectId, requestBody);

        return response.getBatchUid();
    }

    public static void addFileToBatch(JobBatchesApi jobBatchesApi, String projectId, String batchUid, String fileName, String fileUri) throws RestApiRuntimeException, FileNotFoundException
    {
        StreamFileUploadPTO requestBody = StreamFileUploadPTO.builder()
                .fileUri(fileUri)
                .fileType(FILE_TYPE)
                .localeIdsToAuthorize(LOCALES)
                .file(new FileInputStream(fileName))
                .build();

        jobBatchesApi.addFileAsStreamAsync(projectId, batchUid, requestBody);

    }

    public static InputStream downloadTranslatedFile(FilesApi filesApi, String projectId, String fileUri, String localeId) throws RestApiRuntimeException
    {
        DownloadTranslationPTO requestBody = DownloadTranslationPTO.builder()
                .fileUri(fileUri)
                .retrievalType(RetrievalType.PUBLISHED)
                .build();

        return filesApi.downloadTranslatedFile(projectId, localeId, requestBody);

    }

    static String getOutputFileName(String fileName, String locale) {
        // return file names of the form <filename>_<locale>.<extension>
        int dotIndex = fileName.lastIndexOf('.');
        return fileName.substring(0, dotIndex) + "_" + locale + fileName.substring(dotIndex);
    }

}

File: tutorial-1-translating-with-jobs/pythonsdk/process-job.py
import sys
import time


from smartlingApiSdk.Credentials import Credentials
from smartlingApiSdk.api.JobsApi import JobsApi
from smartlingApiSdk.api.JobBatchesV2Api import JobBatchesV2Api
from smartlingApiSdk.api.FilesApi import FilesApi

FILE_LIST = [
    'test-files/test1.json',
    'test-files/test2.json'
]
FILE_TYPE = 'json'
LOCALE_LIST = ['fr-FR']
AUTHORIZE = True 

def main():

    # Credentials expects the following environment variables to be defined:
    # SL_USER_IDENTIFIER, SL_USER_SECRET, SL_PROJECT_ID
    creds = Credentials()

    # instantiate API clients
    jobs_api = JobsApi(creds.MY_USER_IDENTIFIER, creds.MY_USER_SECRET, creds.MY_PROJECT_ID)
    job_batches_api = JobBatchesV2Api(creds.MY_USER_IDENTIFIER, creds.MY_USER_SECRET, creds.MY_PROJECT_ID)
    files_api = FilesApi(creds.MY_USER_IDENTIFIER, creds.MY_USER_SECRET, creds.MY_PROJECT_ID)
    
    # create job
    print('Creating job...')
    job_name = 'Test Job ' + str(int(time.time()))
    resp, status = jobs_api.addJob(job_name)
    if status != 200:
        print(status, resp)
        sys.exit()
    job_uid = resp.data.translationJobUid
    print('Job created.')


    # create job batch
    print('Create job batch...')
    # include job_uid in file URI to avoid confusion during testing
    uri_list = [job_uid + '/' + file_name for file_name in FILE_LIST]
    resp, status = job_batches_api.createJobBatchV2(authorize=AUTHORIZE, 
                                                    translationJobUid=job_uid, 
                                                    fileUris=uri_list)
    if status != 200:
        print(status, resp)
        sys.exit()
    batch_uid = resp.data.batchUid
    print('Job batch created.')


    # upload files to job batch    
    print('Adding files to batch...')
    for file_name in FILE_LIST:
        uri = job_uid + '/' + file_name
        resp, status = job_batches_api.uploadFileToJobBatchV2(batchUid=batch_uid, 
                                                              file=file_name, 
                                                              fileUri=uri, 
                                                              fileType=FILE_TYPE, 
                                                              authorize=AUTHORIZE, 
                                                              localeIdsToAuthorize=LOCALE_LIST)
        if status != 202:
            print(status, resp)
            sys.exit()
        print('Added file ' + file_name)


    # check job status until complete
    print('Will check job status every 10 seconds until complete. Log in to dashboard to complete translations.')
    while True:
        resp, status = jobs_api.getJobProgress(translationJobUid=job_uid)
        if status != 200:
            print(status, resp)
            sys.exit()
        progress = resp.data.progress
        if progress is not None:
            percent_complete = progress['percentComplete']
        else:
            percent_complete = 0
        print('Job progress: ' + str(percent_complete) + '% complete')
        if percent_complete == 100:
            break
        else:
            time.sleep(10)
    
    
    # download translated files
    for file_name in FILE_LIST:
        for locale in LOCALE_LIST:
            uri = job_uid + '/' + file_name
            resp, status = files_api.downloadTranslatedFileSingleLocale(localeId=locale, fileUri=uri)
            if status != 200:
                print(status, resp)
                sys.exit()
            # insert locale_id before '.json' extension in translated file name
            translated_file_name = file_name[:-5] + '_' + locale + file_name[-5:]
            with open(translated_file_name, 'wb') as f:
                f.write(resp)
                print('Downloaded ' + translated_file_name)


if __name__ == '__main__':
    main()



